proto/tendermint/blockchain/message.go

```diff
@@ -0,0 +1,102 @@
+package blockchain
+
+import (
+	"errors"
+	fmt "fmt"
+
+	proto "github.com/gogo/protobuf/proto"
+)
+
+// Wrap implements the p2p Wrapper interface and wraps a blockchain messages.
+func (m *Message) Wrap(pb proto.Message) error {
+	switch msg := pb.(type) {
+	case *BlockRequest:
+		m.Sum = &Message_BlockRequest{BlockRequest: msg}
+
+	case *BlockResponse:
+		m.Sum = &Message_BlockResponse{BlockResponse: msg}
+
+	case *NoBlockResponse:
+		m.Sum = &Message_NoBlockResponse{NoBlockResponse: msg}
+
+	case *StatusRequest:
+		m.Sum = &Message_StatusRequest{StatusRequest: msg}
+
+	case *StatusResponse:
+		m.Sum = &Message_StatusResponse{StatusResponse: msg}
+
+	default:
+		return fmt.Errorf("unknown message: %T", msg)
+	}
+
+	return nil
+}
+
+// Unwrap implements the p2p Wrapper interface and unwraps a wrapped blockchain
+// message.
+func (m *Message) Unwrap() (proto.Message, error) {
+	switch msg := m.Sum.(type) {
+	case *Message_BlockRequest:
+		return m.GetBlockRequest(), nil
+
+	case *Message_BlockResponse:
+		return m.GetBlockResponse(), nil
+
+	case *Message_NoBlockResponse:
+		return m.GetNoBlockResponse(), nil
+
+	case *Message_StatusRequest:
+		return m.GetStatusRequest(), nil
+
+	case *Message_StatusResponse:
+		return m.GetStatusResponse(), nil
+
+	default:
+		return nil, fmt.Errorf("unknown message: %T", msg)
+	}
+}
+
+// Validate validates the message returning an error upon failure.
+func (m *Message) Validate() error {
+	if m == nil {
+		return errors.New("message cannot be nil")
+	}
+
+	switch msg := m.Sum.(type) {
+	case *Message_BlockRequest:
+		if m.GetBlockRequest().Height < 0 {
```

Comment:
Why do we call `GetBlockRequest()` here? We already have access to the message as the appropriate type via `msg`, so we can simply do `msg.Height`. The same goes for the ones below.

Reply:
No, it would be `m.GetBlockRequest().Height` vs `msg.BlockRequest.Height`. There isn't really a difference.

Reply:
I guess the latter saves us a few columns of code lol. I don't have a strong preference on this. Lmk if you do.

Reply:
Ah, right. Protobuf doesn't really fit too well with the Go type system. :/ Either is fine.